ftdc/cmd/parser.go

```diff
@@ -370,24 +433,109 @@ func (gpw *gnuplotWriter) CompileAndClose() string {
+		if graphInfo.minVal == graphInfo.maxVal {
+			graphInfo.maxVal = graphInfo.minVal + 1
+		}
+
+		// Set the lower/upper limits on the Y-axis for the graph. We use 2* maxVal to allow extra
+		// headroom for writing out the legend.
+		writelnf(gnuFile, "set yrange [%v:%v]", graphInfo.minVal, 2*float64(graphInfo.maxVal))
+
+		//nolint
+		// We write the plot line without a trailing newline. In case we want to add the vertical
+		// lines. An example output for the following writes might be:
+		// plot '/tmp/ftdc_parser1500397090/3060093295' using 1:2 with lines linestyle 7 lw 4 title 'rdk-internal:service:web/builtin.ResStats.RPCServer.WebRTCGrpcStats.CallTicketsAvailable',\
+		// 		'/tmp/ftdc_parser1500397090/vert-0.txt' using 1:2 with lines linestyle 6 lw 4 title '2025-02-22 21:13:00 +0000 UTC',\
+		// 		'/tmp/ftdc_parser1500397090/vert-1.txt' using 1:2 with lines linestyle 6 lw 4 title '2025-02-22 21:13:40 +0000 UTC'
+		//
+		//
+		// linestyle 7 is red, 6 is blue, lw is line-width (or weight) -- makes it thicker. The
+		// title is what's used in the legend.
+		writef(gnuFile, "plot '%v' using 1:2 with lines linestyle 7 lw 4 title '%v'", graphInfo.file.Name(), strings.ReplaceAll(metricName, "_", "\\_"))
+
+		// "vertical lines" for events are rendered as another set of data points for a
+		// `plot`. Because the vertical lines are at the same x-value/time for each graph, we can
+		// re-use the same file at a pre-determined name. These files will be written out next after
+		// we've accumulated all of the min/max Y values.
+		for idx, vertLineX := range gpw.options.vertLinesAtSeconds {
+			writeln(gnuFile, ",\\")
+			writef(gnuFile,
+				"\t'%v' using 1:2 with lines linestyle 6 lw 4 title '%v'",
+				filepath.Join(gpw.tempdir, fmt.Sprintf("vert-%d.txt", idx)),
+				time.Unix(vertLineX, 0).UTC())
+		}
+
+		// The trailing newline for the above calls to write out a single plot.
+		writeln(gnuFile, "")
+
+		utils.UncheckedErrorFunc(graphInfo.file.Close)
+	}
+	if allZeroesHidden > 0 {
+		nolintPrintln("Hid metrics that only had 0s for data. Cnt:", allZeroesHidden)
+		// Dan: perhaps add a command to disable this. E.g:
+		//   nolintPrintln("Use `set show-all-zeroes 1` to show them.")
+	}
+
+	// Actually write out the `vert-<number>.txt` plots.
+	for idx, vertLineX := range gpw.options.vertLinesAtSeconds {
+		vertFile, err := os.Create(filepath.Join(gpw.tempdir, fmt.Sprintf("vert-%v.txt", idx)))
+		defer vertFile.Close()
+		if err != nil {
+			panic(err)
+		}
+
+		// For same reason, trying to graph two points (time, -inf) -> (time, +inf) does not
+		// render. So we instead write (time, -inf) -> (time, 0) -> (time, +inf). Which does work...
+		writelnf(vertFile, "%v %d", vertLineX, minY)
+		writelnf(vertFile, "%v 0", vertLineX)
+		writelnf(vertFile, "%v %d", vertLineX, maxY)
 	}
 
 	return gnuFile.Name()
 }
 
+func parseStringAsTime(inp string) (time.Time, error) {
+	goTime, err := time.Parse("2006-01-02T15:04:05", inp)
```

Comment:
I assume this format string is indicated to users somewhere in a `help` command or something?

Reply:
Yep: https://github.com/viamrobotics/rdk/pull/4813/files#diff-a6de85d19062a282c0d7f2c427b26f6598fcf073ef34752a97ec02b053a62f18R591

Also the error line just below also gives an examples: https://github.com/viamrobotics/rdk/pull/4813/files#diff-a6de85d19062a282c0d7f2c427b26f6598fcf073ef34752a97ec02b053a62f18R532

Reply:
Nice! Thanks.